### PR TITLE
Refactor the readers to have a single reader class

### PR DESCRIPTION
**Pull request by mhutchinson**
The strategy for choosing the next leaf to read is now dictated by a
function. This allows much of the logic to be simplified and the getLeaf
function can now become an method of the reader, which avoids needing to
pass in as many parameters.

### CODE REVIEW

File: hammer/clients.go [hunk 1]
```diff
@@ -74,107 +75,34 @@ func (r *RandomLeafReader) Run(ctx context.Context) {
...
-		case <-r.throttle:
-		}
-		klog.V(2).Infof("FullLogReader getting %d", r.current)
-		_, err := getLeaf(ctx, r.f, r.current, r.tracker.LatestConsistent.Size, r.bundleSize)
+		klog.V(2).Infof("LeafReader getting %d", i)
+		_, err := r.getLeaf(ctx, i, size)
 		if err != nil {
-			r.errchan <- fmt.Errorf("Failed to get next leaf: %v", err)
-			continue
+			r.errchan <- fmt.Errorf("Failed to get random leaf: %v", err)
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply hammer/clients.go]: Done

File: hammer/clients.go [hunk 2]
```diff
@@ -74,107 +75,34 @@ func (r *RandomLeafReader) Run(ctx context.Context) {
...
-		if r.current >= r.tracker.LatestConsistent.Size {
-			klog.V(2).Infof("FullLogReader has consumed whole log of size %d. Sleeping.", r.tracker.LatestConsistent.Size)
-			// Sleep a bit and then try again
-			select {
-			case <-ctx.Done(): //context cancelled
-				return
-			case <-time.After(2 * time.Second): //timeout
-			}
+		i := r.next(size)
+		if i == size {
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply hammer/clients.go]: Done

File: hammer/clients.go [hunk 3]
```diff
@@ -27,41 +27,42 @@ import (
...
+		next:       next,
 		bundleSize: bundleSize,
 		throttle:   throttle,
 		errchan:    errchan,
 	}
 }
 
-// RandomLeafReader reads random leaves across the tree.
-type RandomLeafReader struct {
+// LeafReader reads random leaves across the tree.
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply hammer/clients.go]: Done

File: hammer/clients.go [hunk 4]
```diff
@@ -27,41 +27,42 @@ import (
...
 
 	"github.com/transparency-dev/serverless-log/api/layout"
 	"github.com/transparency-dev/serverless-log/client"
 	"k8s.io/klog/v2"
 )
 
-// NewRandomLeafReader creates a RandomLeafReader.
-func NewRandomLeafReader(tracker *client.LogStateTracker, f client.Fetcher, bundleSize int, throttle <-chan bool, errchan chan<- error) *RandomLeafReader {
+// NewLeafReader creates a LeafReader.
+func NewLeafReader(tracker *client.LogStateTracker, f client.Fetcher, next func(uint64) uint64, bundleSize int, throttle <-chan bool, errchan chan<- error) *LeafReader {
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply hammer/clients.go:39]: Done

File: hammer/clients.go [hunk 5]
```diff
@@ -76,13 +78,13 @@ func (r *LeafReader) Run(ctx context.Context) {
...
 		i := r.next(size)
-		if i == size {
+		if i >= size {
 			continue
 		}
 		klog.V(2).Infof("LeafReader getting %d", i)
 		_, err := r.getLeaf(ctx, i, size)
 		if err != nil {
-			r.errchan <- fmt.Errorf("Failed to get random leaf: %v", err)
+			r.errchan <- fmt.Errorf("Failed to get leaf: %v", err)
```
Comment (AlCutter) [inline hammer/clients.go:87]: nit: "failed ..."

**Review by AlCutter — APPROVED** [review]

File: hammer/clients.go (hunk 5 above)
Comment (mhutchinson) [reply hammer/clients.go]: done

**Pull request merged**